File: 3939-process-string-with-special-operations-ii/process-string-with-special-operations-ii.py
class Solution:
    def processStr(self, s: str, k: int) -> str:
        # n=len(s)

        len=0
        for ch in s:
            if ch == '*':
                len -= 1
            elif ch == '#':
                len *= 2
            elif ch == '%':
                continue
            else:
                len+=1
            if len<0:
                len=0

        k+=1
        if k>len:
            return '.'

        s=s[::-1]
        for ch in s:
            if ch == '*':
                len+=1
            elif ch == '#':
                if k>(len//2):
                    k -= (len//2)
                len//=2
            elif ch == '%':
                k=(len-k+1)
            else:
                if k == len:
                    return ch
                len-=1
        
        return '.'

File: 3939-process-string-with-special-operations-ii/process-string-with-special-operations-ii.java
class Solution{
    public char processStr(String s, long k){
        int n=s.length();

        long len=0;
        for(char ch: s.toCharArray()){
            if(ch == '*'){
                len -= 1;
            }
            else if(ch == '#'){
                len *= 2;
            }
            else if(ch == '%'){
                continue;
            }
            else{
                len++;
            }
            if(len<0)
                len=0;
        }

        k++;
        if(k>len)
            return '.';

        for(int i=s.length()-1; i>=0; i--){
            if(s.charAt(i) == '*'){
                len++;
            }
            else if(s.charAt(i) == '#'){
                if(k>(len/2)){
                    k -= (len/2);
                }
                len/=2;
            }
            else if(s.charAt(i) == '%'){
                k=(len-k+1);
            }
            else{
                if(k == len)
                    return s.charAt(i);
                len--;
            }
        }

        return '.';

    }
}